packager/packager.cc

```diff
@@ -502,6 +509,41 @@ ast::ExpressionPtr prependPackageScope(const core::GlobalState &gs, ast::Express
     return scope;
 }
 
+uint8_t verifyLayer(core::MutableContext ctx, core::NameRef fun, ast::ExpressionPtr &expr) {
+    auto target = ast::cast_tree<ast::Literal>(expr);
+    const auto &packageDB = ctx.state.packageDB();
+    if ((target == nullptr && packageDB.getLayerCount() > 0) || !target->isString(ctx.state)) {
+        if (auto e = ctx.beginError(expr.loc(), core::errors::Packager::InvalidLayer)) {
+            string layers;
+            layers.append(packageDB.getLayerName(0));
+            for (int i = 1; i < packageDB.getLayerCount(); i++) {
+                layers.append(", ");
+                layers.append(packageDB.getLayerName(i));
+            }
+            e.setHeader("Argument to `{}` must be a valid layer name ({})", fun.show(ctx), layers);
+        }
+        return 0;
+    }
+
+    string layerName = target->asString(ctx.state).toString(ctx.state);
```

Comment:
storing things as NameRefs exclusively will let you drop the `toString`, so that `layerName` here can be `NameRef`.

Reply:
++